src/main/scala/com/fulcrumgenomics/util/zip/Zipper.scala

```diff
@@ -0,0 +1,840 @@
+  val sizeInBytes: Long = {
+    SequentialBlocks.sizeInBytes(true) + SequentialBlocks.sizeInBytes(false)
+  }
+
+  /** The default block length in bytes of uncompressed data. */
+  val DefaultUncompressedFixedBlockLength = DEFAULT_UNCOMPRESSED_BLOCK_SIZE
+
+  /** The maximum block length in bytes of compressed data (ignores the GZIP header). */
+  val MaxCompressedFixedBlockLengthNoHeader = MAX_COMPRESSED_BLOCK_SIZE - BLOCK_HEADER_LENGTH
+
+  /** The maximum block length in bytes of compressed data (with the GZIP header and footer). */
+  val MaxCompressedFixedBlockLength = MAX_COMPRESSED_BLOCK_SIZE - 2 // subtract two for the BGZF extra field
+
+  /** Companion definitions and methods for sequential blocks. */
+  private[zip] object SequentialBlocks {
+
+    /** The default # of blocks in a sequential block. */
+    private[zip] var DefaultCapacity: Int = 128
+
+    /** Returns the approximate size in bytes for Blocks. */
+    def sizeInBytes(compress: Boolean, capacity: Int = DefaultCapacity): Long = {
+      if (compress) MaxCompressedFixedBlockLength * capacity
+      else DefaultUncompressedFixedBlockLength * capacity
+    }
+  }
+
+  /**
+    * Stores multiple blocks to compress or decompress.  This facilitates more efficient reading and writing
+    * of the data, since more data can be read or written at any one time.  It also allows for multiple blocks to
+    * be compressed or decompressed by a single consumer.
+    *
+    * @param capacity the capacity of fixed-size blocks.
+    * @param compress true if the data will be compressed, false if it is to be decompressed.
+    */
+  private class SequentialBlocks private(capacity: Int, compress: Boolean) {
+    def this(compress: Boolean) = {
+      this(capacity=SequentialBlocks.DefaultCapacity, compress=compress)
+    }
+
+    /** The buffer storing the data. */
+    val byteBuffer = {
+      if (compress) ByteBuffer.allocate(MaxCompressedFixedBlockLength * capacity)
+      else ByteBuffer.allocate(DefaultUncompressedFixedBlockLength * capacity)
+    }
+    byteBuffer.order(ByteOrder.LITTLE_ENDIAN)
+
+    /** The number of blocks stored. */
+    var numStoredBlocks: Int = 0
+
+    /** The size of each stored block. */
+    var blockSizes = new ListBuffer[Int]() // NB: for uncompressed data, only the last size can be different
+
+    /** Gets the underlyng array of bytes. */
+    def array: Array[Byte] = this.byteBuffer.array()
+
+    /** Clears the underlying bytes, so this object can be re-used. */
+    def clear(): Unit = {
+      this.byteBuffer.clear()
+      blockSizes.clear()
+      numStoredBlocks = 0
+    }
+  }
+}
+
+/** Stores the input data to be processed, and the output data after processing.
+  *
+  * @param blockId the block identifier, typically the index of the block as it was read.
+  * @param compress true if the input data will be compressed, false otherwise.
+  */
+private class Blocks(var blockId: Long, compress: Boolean) extends Ordered[Blocks] {
```

Comment:
The names of these classes make me think that `SequentialBlocks` will contain many `Block[s]`, not the other way around.